internal/controller/domain_controller.go

```
@@ -329,6 +353,386 @@ func defaultHTTPGet(ctx context.Context, url string) ([]byte, *http.Response, er
+		}
+	} else {
+		// Non-apex: find delegated NS for the exact name; if none, fall back to apex NS
+		if _, nsHosts, delegated := r.delegatedZoneNS(ctxLookup, d.Spec.DomainName, apex); delegated && len(nsHosts) > 0 {
+			logger.Info("subdomain delegation detected; using delegated nameservers",
+				"zoneApex", d.Spec.DomainName, "nsCount", len(nsHosts))
+			for _, h := range nsHosts {
+				st.Nameservers = append(st.Nameservers, networkingv1alpha.Nameserver{Hostname: h})
+			}
+		} else {
+			// fallback to apex NS from RDAP
+			for _, ns := range rdapDomain.Nameservers {
+				st.Nameservers = append(st.Nameservers, networkingv1alpha.Nameserver{Hostname: ns.LDHName})
+			}
+		}
+	}
+
+	// Enrich top-level Nameservers with per-IP registrant
+	for i := range st.Nameservers {
+		host := st.Nameservers[i].Hostname
+		ipAddrs, err := r.lookupIP(ctxLookup, host)
+		if err != nil {
+			logger.Error(err, "error looking up IP addresses for nameserver", "host", host)
+			continue
+		}
+		st.Nameservers[i].IPs = nil
+		for _, ipAddr := range ipAddrs {
+			ip := ipAddr.IP.String()
+			nsIP := networkingv1alpha.NameserverIP{Address: ip}
+			if who := r.lookupRegistrantNameForIP(ctxLookup, ip); who != "" {
+				nsIP.RegistrantName = who
+			}
+			st.Nameservers[i].IPs = append(st.Nameservers[i].IPs, nsIP)
+		}
+	}
+
+	reg.Source = "rdap"
+	st.Registration = reg
+
+	// Schedule next refresh (with jitter)
+	interval := r.Config.DomainRegistration.RefreshInterval.Duration
+	next := now.Add(wait.Jitter(interval, r.Config.DomainRegistration.JitterMaxFactor))
+	st.Registration.NextRegistrationAttempt = metav1.NewTime(next)
+
+	return next
+}
+
+// mapRDAPDomainToRegistration maps a raw RDAP domain into our Registration model.
+func (r *DomainReconciler) mapRDAPDomainToRegistration(d *rdap.Domain) *networkingv1alpha.Registration {
+	reg := &networkingv1alpha.Registration{}
+	if d == nil {
+		return reg
+	}
```

I'm not seeing a scenario where `d` would be nil. If we want to protect against that changing in the future, would it make more sense to return an error instead of an empty registration struct?